truffle/com.oracle.truffle.object.dsl.processor/src/com/oracle/truffle/object/dsl/processor/layout/LayoutGenerator.java

```diff
@@ -0,0 +1,839 @@
+            if (property.hasObjectTypeGetter()) {
+                stream.println("    @Override");
+                stream.printf("    public %s %s(ObjectType objectType) {%n", property.getType(), NameUtils.asGetter(property.getName()));
+                stream.printf("        assert is%s(objectType);%n", layout.getName());
+                stream.printf("        return ((%sType) objectType).%s();%n", layout.getName(), NameUtils.asGetter(property.getName()));
+                stream.println("    }");
+                stream.println("    ");
+            }
+
+            if (property.hasFactoryGetter()) {
+                stream.println("    @Override");
+                stream.printf("    public %s %s(DynamicObjectFactory factory) {%n", property.getType(), NameUtils.asGetter(property.getName()));
+                stream.printf("        assert creates%s(factory);%n", layout.getName());
+                stream.printf("        return ((%sType) factory.getShape().getObjectType()).%s();%n", layout.getName(), NameUtils.asGetter(property.getName()));
+                stream.println("    }");
+                stream.println("    ");
+            }
+
+            if (property.hasGetter()) {
+                addUncheckedCastWarning(stream, property);
+                stream.println("    @Override");
+                stream.printf("    public %s %s(DynamicObject object) {%n", property.getType(), NameUtils.asGetter(property.getName()));
+                stream.printf("        assert is%s(object);%n", layout.getName());
+
+                if (property.isShapeProperty()) {
+                    stream.printf("        return getObjectType(object).%s();%n", NameUtils.asGetter(property.getName()));
+                } else {
+                    stream.printf("        assert object.getShape().hasProperty(%s_IDENTIFIER);%n", NameUtils.identifierToConstant(property.getName()));
+                    stream.println("        ");
+
+                    if (property.isVolatile()) {
+                        if (property.getType().getKind() == TypeKind.INT) {
+                            stream.printf("        return ((AtomicInteger) %s_PROPERTY.get(object, true)).get();%n", NameUtils.identifierToConstant(property.getName()));
+                        } else if (property.getType().getKind() == TypeKind.BOOLEAN) {
+                            stream.printf("        return ((AtomicBoolean) %s_PROPERTY.get(object, true)).get();%n", NameUtils.identifierToConstant(property.getName()));
+                        } else {
+                            stream.printf("        return ((AtomicReference<%s>) %s_PROPERTY.get(object, true)).get();%n", property.getType(), NameUtils.identifierToConstant(property.getName()));
+                        }
+                    } else {
+                        stream.printf("        return (%s) %s_PROPERTY.get(object, true);%n", property.getType(), NameUtils.identifierToConstant(property.getName()));
```

Comment:
I presume ideally it would be a boolean data edge from the earlier `is%s(object)` call that the user is presumably doing? But we don't have access to that in the DSL.

Can you tell me about the difference between doing `PROPERTY.get(object, true)`,  `PROPERTY.get(object, false)` and  `PROPERTY.get(object, isFoo(object))`. I think there are some subtleties here that I may not be aware of.